### DISCUSSION TITLE
Firmware version not detected for latest firmwares

### DISCUSSION THREAD

**Discussion by eliboa** [eliboa/NxNandManager#75] [General]
_Actually that's not an issue. NxNandManager must be updated to detect new firmware versions._

### How it works now 
A firmware package contains a specifc title (program): ```0100000000000809 (SystemVersion)```. Materially, this program is binary data stored inside a NCA file ([Nintendo Content Archive](https://switchbrew.org/wiki/NCA)) whose filename is different for each firmware version.    
For example: 
- fce3b0ea366f9c95fe6498b69274b0e7.nca  = SystemVersion title for FW 5.1.0
- 5625cdc21d5f1ca52f6c36ba261505b9.nca = SystemVersion title for FW 10.0.0
- etc.

A [static array](https://github.com/eliboa/NxNandManager/blob/27a2884f6d0d5f14279b6d7a7a3fe49da406989f/NxNandManager/NxStorage.cpp#L68) in source code, manually updated, references each NCA filename and corresponding version.
When the suitable bis key is provided, NxNandManager will scan the content of SYSTEM:/Contents/registered. If a *.nca file inside this folder matches an entry in the array, you get your firmware version. Simple as that.

### How to get the new NCA filename each time the firmware is updated ?
A firmware package is just a bunch of NCAs. All theses files are fully encrypted.
With the right keys, you can decrypt NCA header (first 0xC00 bytes) where the program Id (title Id) is stored. If the program Id is 0100000000000809, you know this is "SystemVersion".

So every time a firmware package is released, with a program like hactool, you can get the SystemVersion's filename by inspecting the package.

### What's the problem?
1) Someone has to decrypt and inspect the content of the firmware package every time a new one is released.
2) Someone has to update NxNandManager's code (static array).
3) Someone has to build & release a new version.
4) End user has to download latest version of NxNM

### Solutions
A) Implement NCA header decryption 
- Resolves problem 1, 2, 3 & 4. 
- For end user, this would require importing new keys.   

B) Request a remote DB to get the latest nca filenames instead of a static array 
- Resolves problem 2, 3, & 4.  
- Quick win, easy to implement but you still need someone or some algorithm to update the DB.

C) ???

### Implementation
Solution B. Will only work with Qt (GUI version).

Remote DB ✔️ https://eliboa.com/switch/nca.php
DB request implementation ✔️ 
Quality tests ✔️ 
Release ✔️

**Comment by tldzyx** [comment]
maybe just one simple config file replace source code(static array) like solutions B, and it's even work without network(in some network limit place).

**Comment by impeeza** [comment]
Sadly the latest successful Action Build has expired, and the new code do not build.

**Reply by tldzyx** [reply]
if there have a dev branch, at least it can be build, then merge to master, it's will better

**Comment by impeeza** [comment]
The latest build can be downloaded from  https://gbatemp.net/download/nx-nand-manager.38831/  that's the version 5.2 commit e8b5ce5 which is the most recent commit what builds from code, posterior commits have errors and can not be build.  once those errors will be corrected you should get the most recent from [Action Menu](https://github.com/eliboa/NxNandManager/actions)

**Reply by tldzyx** [reply]
thank you very much, I got it ^_^
